src/protocol.rs

```diff
@@ -1,843 +0,0 @@
-//! API for protocol functionality such as creating and parsing ODoH queries and responses.
-
-use aes_gcm::aead::{generic_array::GenericArray, AeadInPlace, NewAead};
-use aes_gcm::Aes128Gcm;
-use anyhow::{anyhow, Context, Error, Result};
-use bincode2::LengthOption;
-use hkdf::Hkdf;
-use hpke::{
-    aead::{AeadTag, AesGcm128},
-    kdf::HkdfSha256,
-    kdf::Kdf as KdfTrait,
-    kem::X25519HkdfSha256,
-    kex::KeyExchange,
-    AeadCtxR, Deserializable, EncappedKey, Kem as KemTrait, OpModeR, OpModeS, Serializable,
-};
-use rand::{rngs::StdRng, Rng, SeedableRng};
-use serde::{Deserialize as SerdeDeserialize, Serialize as SerdeSerialize};
-use serde_repr::*;
-use std::{
-    convert::{TryFrom, TryInto},
-    vec,
-};
-
-/// HTTP header required for sending queries and responses
-pub const ODOH_HTTP_HEADER: &str = "application/oblivious-dns-message";
-const LABEL_QUERY: &[u8] = b"odoh query";
-const LABEL_KEY: &[u8] = b"odoh key";
-const LABEL_NONCE: &[u8] = b"odoh nonce";
-const LABEL_KEY_ID: &[u8] = b"odoh key id";
-const LABEL_RESPONSE: &[u8] = b"odoh response";
-/// ODoH version supported by this library
-pub const ODOH_VERSION: u16 = 0xff06;
-
-pub type Kem = X25519HkdfSha256;
-pub type Aead = AesGcm128;
-pub type Kdf = HkdfSha256;
-pub type Kex = <Kem as KemTrait>::Kex;
-const KEM_ID: u16 = 0x0020;
-const KDF_ID: u16 = 0x0001;
-const AEAD_ID: u16 = 0x0001;
-
-/// For the selected KDF: SHA256
-const KDF_OUTPUT_SIZE: usize = 32;
-const AEAD_KEY_SIZE: usize = 16;
-const AEAD_NONCE_SIZE: usize = 12;
-const AEAD_TAG_SIZE: usize = 16;
-/// This is the maximum of `AEAD_KEY_SIZE` and `AEAD_NONCE_SIZE`
-pub const RESPONSE_NONCE_SIZE: usize = 16;
-
-macro_rules! impl_custom_serde {
-    ($name:ident) => {
-        impl Serialize for $name {}
-        impl<'de> Deserialize<'de> for $name {}
-    };
-}
-
-/// Serialize to IETF wireformat that is similar to [XDR](https://tools.ietf.org/html/rfc1014)
-pub trait Serialize {
-    fn to_bytes(&self) -> Result<Vec<u8>>
```

Comment:
If you're trying to not allocate memory here, you can simply do 
```
    fn to_bytes<B: std::io::Write>(&self, buf: &mut B) -> Result<()>
    where
        Self: SerdeSerialize,
    {
        let serialized = bincode2::config()
            .big_endian()
            .array_length(LengthOption::U16)
            .serialize_into(buf, &self)?;
        Ok(serialized)
    }
    ```
  instead of removing `bincode2` entirely. It's much more concise as well thanks to the macro.